src/core/mem_throt.c

```diff
@@ -5,63 +5,122 @@
+    if(cpu()->vcpu->vm->mem_throt.ticket_num_left > 0)
+    {
+        cpu()->vcpu->vm->mem_throt.ticket_num_left--;
+        events_cntr_set(cpu()->vcpu->vm->mem_throt.counter_id, cpu()->vcpu->vm->mem_throt.budget);
+        events_cntr_enable(cpu()->vcpu->vm->mem_throt.counter_id);
 
-    events_clear_cntr_ovs(cpu()->mem_throt.counter_id);
-    events_cntr_disable(cpu()->mem_throt.counter_id);
-    events_cntr_irq_disable(cpu()->mem_throt.counter_id);
-    cpu()->mem_throt.throttled = true;
-    cpu_standby();
+    }
+    else
+    {
+        cpu()->vcpu->vm->mem_throt.throttled = true;  
+        cpu_standby();
+    }
+    //enhace this part
 }
 
-void mem_throt_timer_init(irq_handler_t handler) {
 
+void mem_throt_timer_init(irq_handler_t handler) {
+    // console_printk("Memory Throttling Timer Init on cpu %d\n", cpu()->id);
     timer_define_irq_callback(handler);
-    cpu()->mem_throt.period_counts = timer_init(cpu()->mem_throt.period_us);
+    cpu()->vcpu->vm->mem_throt.period_counts = timer_init(cpu()->vcpu->vm->mem_throt.period_us);
 }
 
 
 void mem_throt_events_init(events_enum event, unsigned long budget, irq_handler_t handler) {
 
-    if ((cpu()->mem_throt.counter_id = events_cntr_alloc()) == ERROR_NO_MORE_EVENT_COUNTERS) {
+    if ((cpu()->vcpu->vm->mem_throt.counter_id = events_cntr_alloc()) == ERROR_NO_MORE_EVENT_COUNTERS) {
         ERROR("No more event counters!");
     }
 
-    events_set_evtyper(cpu()->mem_throt.counter_id, event);
-    events_cntr_set(cpu()->mem_throt.counter_id, budget);
-    events_cntr_set_irq_callback(handler, cpu()->mem_throt.counter_id);
-    events_clear_cntr_ovs(cpu()->mem_throt.counter_id);
+    events_set_evtyper(cpu()->vcpu->vm->mem_throt.counter_id, event);
+    events_cntr_set(cpu()->vcpu->vm->mem_throt.counter_id, budget);
+    events_cntr_set_irq_callback(handler, cpu()->vcpu->vm->mem_throt.counter_id);
+    events_clear_cntr_ovs(cpu()->vcpu->vm->mem_throt.counter_id);
     events_interrupt_enable(cpu()->id);
-    events_cntr_irq_enable(cpu()->mem_throt.counter_id);
+    events_cntr_irq_enable(cpu()->vcpu->vm->mem_throt.counter_id);
     events_enable();
-    events_cntr_enable(cpu()->mem_throt.counter_id);
+    events_cntr_enable(cpu()->vcpu->vm->mem_throt.counter_id);
 }
 
-void mem_throt_init(uint64_t budget, uint64_t period_us) {
-
-    cpu()->mem_throt.throttled = false;
-    cpu()->mem_throt.budget = budget;
-    cpu()->mem_throt.period_us = period_us;
+void mem_throt_init(uint64_t budget, uint64_t period_us, uint64_t ticket_num) {
 
+    if (cpu()->vcpu->vm->master) 
+    {
+        cpu()->vcpu->vm->mem_throt.throttled = false;
+        cpu()->vcpu->vm->mem_throt.period_us = period_us;
+        cpu()->vcpu->vm->mem_throt.ticket_num = ticket_num - cpu()->vcpu->vm->cpu_num;
+        cpu()->vcpu->vm->mem_throt.ticket_num_left = cpu()->vcpu->vm->mem_throt.ticket_num;
+        cpu()->vcpu->vm->mem_throt.budget = budget / cpu()->vcpu->vm->mem_throt.ticket_num;
+    }
```

Comment:
This does not seem to be correct... It looks like that only the ``cpu_master``  is being configure with the budgets. Shouldn't the other CPUs also be configured?

Reply:
In this version, everything is stored inside the vm, so all `tickets` are getting the same part of the budget, ithen, each cpu gets a ticket.

Reply:
I'll add a flag to check if the init has already been done.